### DISCUSSION TITLE
observable.array excess triggers?

### DISCUSSION THREAD

**Discussion by alexandroppolus** [mobxjs/mobx#2811] [General] [answered]
Hi!

mobx@6.1.6

```javascript
const arr = observable.array([0, 0])

autorun(() => {
  console.log('arr[0] === ' + arr[0])
})
// print "arr[0] === 0"

runInAction(() => {
  arr[1] = 1
})
// print "arr[0] === 0" WTF???
```

`autorun` observes to `arr[0]`. Why it triggers on change `arr[1]`?

If using `observable.object({ 0: 0, 1: 0 })`, all good, `autorun` not triggers.

With regards,
Alex

**Comment by mweststrate** [comment, marked as answer]
This works as intended, arrays themselves are considered a single atom.
This is an optimization decision that makes cases like this slower, but
scenarios like looping or shifting items much faster. If per-key
subscriptions are important, use maps insteads.

**Reply by alexandroppolus** [reply]
Ok, thanks.

I also noticed similar behavior in `observable.set`. This are also a single atom?

**Comment by mweststrate** [comment]
correct

**Comment by iChenLei** [comment]
### The difference between three case.

### use `observable.array`
![2021-02-17 17 41 22](https://user-images.githubusercontent.com/14012511/108191885-9ec48880-714e-11eb-9e28-771bd8e0a80d.png)
> This works as intended, arrays themselves are considered a single atom.       
   --weststrate

### use `observable.object`
![2021-02-17 17 50 31](https://user-images.githubusercontent.com/14012511/108191981-b3a11c00-714e-11eb-8fb4-308a9a63e7ec.png)

### use `observable.map`
![截屏2021-02-17 18 32 36](https://user-images.githubusercontent.com/14012511/108192067-c582bf00-714e-11eb-9d05-a3633dd119be.png)

### summary
When you use `observable.array`, the `autorun` track `array` itself, so whatever you do for `array`, the `autorun` will be fire.
`observable.object` or `observable.map` will create a `ObservableValue` for `Set` item or `Object` property , `autorun` only track the special variable's change. More detail is in the mobx source code.
